keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastings.java

```diff
@@ -55,14 +57,16 @@ public static NetworkSamples getPosteriorSamples(final BayesNet bayesNet,
                                  final double logPOld,
                                  final Set<Vertex<?>> affectedVertices,
                                  final double T,
+                                 final Map<String, Map<String, Integer>> setAndCascadeCache,
                                  final Random random) {
 
         final double affectedVerticesLogPOld = sumLogP(affectedVertices);
 
         final T oldValue = chosenVertex.getValue();
         final T proposedValue = chosenVertex.sample();
 
-        chosenVertex.setAndCascade(proposedValue);
+        Map<String, Integer> cascadeCache = setAndCascadeCache.computeIfAbsent(chosenVertex.getId(), (id) -> chosenVertex.exploreSetting());
```

Comment:
My gut instinct is that we can do something a bit sneaky here and avoid having to `exploreSetting` for each Vertex. I see this as being quite a large compute if we have a load of vertices? What about...

1. `exploreSetting` for a Vertex, A, and store it in the map
2. `exploreSetting` for a second Vertex, B, that is `upstream` of A. (B is a parent of A). When we are exploring and find a Vertex that we have already computed, in this case A, can we take the already computed values of how many times vertices will be visited and just add them on to our exploration from B?
3. Add Vertex B's `exploredSetting` to the map 
3. `exploreSetting` for a third Vertex, C, that runs in to B. Use this saved propagation count...
4. etc

TLRD; Sort of like a cached routing problem. Once we hit a vertex we've already calculated the propagation count for, just use the values again as they will be the same from that point as previously computed?